kubeadm: graduate the control-plane phase

neolit123 opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

kubeadm: move the controlplane phase to phase runner.

Follows the example of recently the implemented preflight phase.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Refs https://github.com/kubernetes/kubeadm/issues/1163

**Special notes for your reviewer**:

outdated info:
<details>
@fabriziopandini the sub-phases of controlplane need to add flags.
PTAL at how i did it, but i don't like it that much.

- first i had to pass the internal config to the function NewControlPlanePhase() in NewCmdInit()
- i had to add Flags() to the Phase structs
- these flags are combined with the command flags.
</details>

side note: removed controlplane_test.go as it feels redundant now.

**Does this PR introduce a user-facing change?**:
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
NONE
```

/assign @fabriziopandini 
@kubernetes/sig-cluster-lifecycle-pr-reviews 
/kind feature
/area kubeadm

depends on:
https://github.com/kubernetes/kubernetes/pull/70331
https://github.com/kubernetes/kubernetes/pull/70098

neolit123 committed:
kubeadm: move the control-plane phase to the phase runner

Remove custom flags that were previously available per
sub-phase. Rely on the config passed to 'kubeadm init' for that.

Remove redundant functions in manifests.go.

Move the audit policy settings to the control plane phase (under
API server).

neolit123 commented:
/hold
(bots not working now).

but this PR depends on a feature PR about the kubeadm  kubeconfig phase.

fabriziopandini commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -163,14 +162,19 @@ func NewCmdInit(out io.Writer) *cobra.Command {
 	// initialize the workflow runner with the list of phases
 	initRunner.AppendPhase(phases.NewPreflightMasterPhase())
+	// TODO: a couple of phases should go here.
+	initRunner.AppendPhase(phases.NewControlPlanePhase(initData.cfg))
```
InitData are run time data; it confuses me to get those data passed when cmd are created (and I'm not 100% sure this will behave the same)

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -18,185 +18,155 @@ package phases
-			cmd.Flags().Var(utilflag.NewMapStringString(&cfg.SchedulerExtraArgs), "scheduler-extra-args", "A set of extra flags to pass to the Scheduler or override default ones in form of <flagname>=<value>")
+		switch properties.name {
+		case phaseAPIServer:
+			phase.Flags().Var(utilflag.NewMapStringString(&cfg.APIServerExtraArgs), "apiserver-extra-args", "A set of extra flags to pass to the API Server or override default ones in form of <flagname>=<value>")
```
Those flags should be dropped

fabriziopandini replied:
By dropping those flags, you don't need cfg anymore

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -18,185 +18,155 @@ package phases
+		Name:  "control-plane",
 		Short: "Generates all static Pod manifest files necessary to establish the control plane",
 		Long:  cmdutil.MacroCommandLongDescription,
+		Run:   runControlPlane,
```
If this phase is only a container for sub phases you could omit Run

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -18,185 +18,155 @@ package phases
 	manifestPath := kubeadmconstants.GetStaticPodDirectory()
-	cmd.AddCommand(getControlPlaneSubCommands(manifestPath, "")...)
-	return cmd
+	subPhases := getControlPlaneSubPhases(cfg, manifestPath)
```
If you check at the PR I Iinked, I'm proposing a slightly different approach for doing this.
WDYT?

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -18,185 +18,155 @@ package phases
 		long     string
 		examples string
-		cmdFunc  func(outDir string, cfg *kubeadmapi.InitConfiguration) error
+		cmdFunc  func(outDir string, cfg *kubeadm.InitConfiguration) error
```
If you check to PR I linked, I'm proposing an approach that will allow you to get rid of all the Create* wrappers in favor of createStaticPodFiles

fabriziopandini reviewed (changes requested):
@neolit123 thanks for sharing this WIP.
I don't think we need flags, because flags of phases should be (<)= of flags of the command they belong to (this was discussed in the KEP and if I remember well also in the PR for the phase runner.

In this case, IMO we should drop extra args flags in favor of the config file.

With regards to the implementation, I think there is room for a little bit more of refactoring/clean-up; see e.g. https://github.com/kubernetes/kubernetes/pull/70098 last two commits & comments in line

neolit123 commented:
@fabriziopandini 
i think i've addressed all comments.
removing the flags for sub-phases made things much cleaner.

this PR also now moves the audit policy in the API server sub-phase. PTAL.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -466,31 +465,6 @@ func runInit(i *initData, out io.Writer) error {
-
-	// Temporarily set cfg.CertificatesDir to the "real value" when writing controlplane manifests
-	// This is needed for writing the right kind of manifests
-	i.cfg.CertificatesDir = realCertsDir
```
this code is bit difficult to tackle with phases as is.

it first does a:
`realCertsDir := i.cfg.CertificatesDir`
on line 446 it does a `i.cfg.CertificatesDir = certsDirToWriteTo`
and here it goes back to`i.cfg.CertificatesDir = realCertsDir`

i think my change is breaking here; i need to verify that.

comments appreciated.

neolit123 replied:
ok, understood.

fabriziopandini reviewed (changes requested):
@neolit123. Well done!
You can find inline the answer to your question + some nits
IMO you can now remove WIP now so we can get more eyes on this.

neolit123 commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -443,31 +443,10 @@ func runInit(i *initData, out io.Writer) error {
-
 	// Temporarily set cfg.CertificatesDir to the "real value" when writing controlplane manifests
 	// This is needed for writing the right kind of manifests
 	i.cfg.CertificatesDir = realCertsDir
```
reading the code here it seems to me that in the case of etcd and control plane the dry run folder is not used and instead the real (value from config) value is used.

neolit123 commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -17,186 +17,111 @@ limitations under the License.
-		if err := features.ValidateVersion(features.InitFeatureGates, internalcfg.FeatureGates, internalcfg.KubernetesVersion); err != nil {
-			kubeadmutil.CheckErr(err)
+		fmt.Printf("[control-plane] creating static Pod manifest for %s", component)
+		if err := controlplane.CreateStaticPodFiles(data.ManifestDir(), cfg, component); err != nil {
```
seems to me that the original code in runInit() used the same folder for dry-run and without dry-run in the case of etcd and api server.

here by using ManifestDir() we now respect the dry-run flag.

fabriziopandini replied:
I don't have history here, but what you are proposing seems ok to me.

neolit123 commented:
@fabriziopandini manged to update the PR today.
PTAL.

fabriziopandini commented:
/approve
@neolit123 IMO now we are ready (but please keep hold until other phases gets merged)
The problem in dry run was fixed in kubelet-start phase, so if you rebase your PR you will get integration test passing. It seems you need also tu run upgrade-generared -docs.sh to pass kubernetes verify

neolit123 commented:
> @neolit123 IMO now we are ready (but please keep hold until other phases gets merged)
The problem in dry run was fixed in kubelet-start phase, so if you rebase your PR you will get integration test passing.

sounds good.

> It seems you need also tu run upgrade-generared -docs.sh to pass kubernetes verify

sadly my PR that nukes the placeholder docs is blocked by a "riddle" from a maintainer, with not further clarifications.

timothysc commented:
/assign @timothysc

neolit123 committed:
kubeadm: autogenerated docs

neolit123 commented on cmd/kubeadm/app/cmd/phases/kubeconfig_test.go:
```diff
@@ -46,6 +46,8 @@ func TestKubeConfigCSubCommandsHasFlags(t *testing.T) {
 	cmdtestutil.AssertSubCommandHasFlags(t, []*cobra.Command{cmd}, "user", flags...)
 }
 
+const phaseTestK8sVersion = "v1.11.0"
```
this seems to be a constant that we are passing around we probably should just clean it up later.
_test.go files depend on it.

neolit123 commented:
@fabriziopandini updated the PR, fixed the integration issue too.

output with -v=0 is:
```
[control-plane] creating static Pod manifest for kube-apiserver
[control-plane] creating static Pod manifest for kube-controller-manager
[control-plane] creating static Pod manifest for kube-scheduler
````
-v=1 shows the path where files are written.

will re-visit again tomorrow if needed.

neolit123 commented:
/retest

fabriziopandini commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -436,31 +435,10 @@ func runInit(i *initData, out io.Writer) error {
-
 	// Temporarily set cfg.CertificatesDir to the "real value" when writing controlplane manifests
 	// This is needed for writing the right kind of manifests
 	i.cfg.CertificatesDir = realCertsDir
```
@neolit123 If I'm not wrong now it possible to nuke set/reset of `certificateDir` from `runInit`

neolit123 replied:
i'm removing it in the etcd phase PR but we can do it here.

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -17,186 +17,111 @@ limitations under the License.
+			newControlPlaneSubPhase(kubeadmconstants.KubeAPIServer),
+			newControlPlaneSubPhase(kubeadmconstants.KubeControllerManager),
+			newControlPlaneSubPhase(kubeadmconstants.KubeScheduler),
+		},
```
in certs and kubeconfig I have added a run method to the top level phase that prints the folder where stuff are going to be saved. What do you think about doing the same here?

neolit123 replied:
SGTM

fabriziopandini commented on cmd/kubeadm/app/cmd/phases/controlplane.go:
```diff
@@ -17,186 +17,111 @@ limitations under the License.
-
-		if err := features.ValidateVersion(features.InitFeatureGates, internalcfg.FeatureGates, internalcfg.KubernetesVersion); err != nil {
-			kubeadmutil.CheckErr(err)
+		fmt.Printf("[control-plane] creating static Pod manifest for %s\n", component)
```
%s/%q for consistency with other phases

fabriziopandini reviewed (commented):
@neolit123 ready to go!
Some minor that can be fixed now or later when we will do a final review on UX

/approve
/lgtm

neolit123 commented:
@fabriziopandini addressed latest feedback.

fabriziopandini reviewed (commented):
@neolit123 Thanks
/approve
/lgtm

neolit123 commented:
@fabriziopandini 
will rebase the etcd phase later today.

timothysc reviewed (approved):
/lgtm 
/approve

neolit123 commented:
/hold cancel
